### PR TITLE
Draft of a portability chapter explaining the HAL

### PR DESCRIPTION
**Pull request by therealprof**
This is a first draft of an explanation of what a HAL is and what the puzzle pieces are in embedded-hal use.

r? @rust-embedded/resources

### CODE REVIEW

**Comment by therealprof** [comment]
Related to #6

File: src/portability/portability.md [hunk 1]
```diff
@@ -1,3 +1,56 @@
...
+
+[Hardware Abstraction Layer]: https://en.wikipedia.org/wiki/Hardware_abstraction
+
+Embedded systems are a bit special in this regard since we typically do not have operating systems and user installable software but firmware images which are compiled as a whole as well as a number of other constraints. So while the traditional approach as defined by Wikipedia could potentially work it is likely not the most productive approach to ensure portability.
+
+How do we do this in Rust? Enter **embedded-hal**...
+
+## What is embedded-hal?
+
+In a nutshell it is a set of traits which define implementation contracts between **HAL implementations**, **drivers** and **applications (or firmwares)**. Those contracts include both capabilities (i.e. if a trait is implemented for a certain type, it is easy to see that a **HAL implementation** provides a certain capability) and methods (i.e. if you can construct a type implementing a trait it is guarenteed that you have the methods specified in the trait available).
```
Comment (jamesmunns) [inline src/portability/portability.md]: s/guarenteed/guaranteed

Comment (jamesmunns) [inline src/portability/portability.md]: Suggestion, avoid phrases like "it is easy to see", as it might not be easy for all people

Comment (jamesmunns) [inline src/portability/portability.md]: s/guarenteed/guaranteed

File: src/portability/portability.md [hunk 2]
```diff
@@ -1,3 +1,56 @@
...
+* GPIO (input and output pins)
+* Serial communication
+* I2C
+* SPI
+* Timers/Countdowns
+* Analog Digital Conversion
+
+## Users of the embedded-hal
+
+As said above there're three main users of the HAL:
```
Comment (jamesmunns) [inline src/portability/portability.md]: "there're" => "there are"

File: src/portability/portability.md [hunk 3]
```diff
@@ -1,3 +1,56 @@
...
+
+### HAL implementation
+
+A HAL implementation provides the interfacing between the hardware and and the users of the HAL traits. Typical implementations consist of three parts:
+* One or more hardware specific types
+* Functions to create and intialise such a type, often providing various configuration options (speed, operation mode, use pins, etc.)
```
Comment (jamesmunns) [inline src/portability/portability.md]: Not sure if we are standardizing on british vs american english, items in this doc include:
* initialise/initialize
* initialised/initialized
* initialisation/initialization
* flavours/flavors

File: src/portability/portability.md [hunk 4]
```diff
@@ -1,3 +1,56 @@
...
+
+Such a **HAL implementation** can come in various flavours:
+* Via low-level hardware access, e.g. via registers
+* Via operating system, e.g. by using the `sysfs` under Linux
+* Via adapter, e.g. a mock of types for unit testing
+* Via driver for hardware adapters, e.g. I2C multiplexer or GPIO expander
+
+### Driver
+
+A driver implements a set of custom functionality for an internal or external component, connected to a peripheral implementating the embedded-hal traits. Typical examples for such drivers include various sensors (temperatur, magnetometer, accelerometer, light), display devices (LED arays, LCD displays) and actors (motors, relais).
```
Comment (jamesmunns) [inline src/portability/portability.md]: s/implementating/implementing

Comment (jamesmunns) [inline src/portability/portability.md]: s/temperatur/temperature

Comment (jamesmunns) [inline src/portability/portability.md]: Also
* LED arays => LED arrays
* relais => relays

File: src/portability/portability.md [hunk 5]
```diff
@@ -1,3 +1,56 @@
...
+
+A driver has to be initialised with an instance of type that implements a certain `trait` of the embedded-hal which is ensured via trait bound and provides its own type instance with a custom set of methods allowing to interact with the driven device.
+
+### Application
+
+The application binds the various parts together and ensures that the desired functionality is achieved. When porting between different systems, this is the part which requires the most adaptation efforts, since the application needs to correctly intialise the real hardware via the HAL implementation and the initialisation of different hardware differs, sometimes drastically so. Also the user choice often plays a big role, since components can be physically connected to different terminals, hardware busses sometimes need external hardware to match the configuration or there're different tradeoffs to be made in the use of internal peripherals (e.g. multiple timers with different capabilities are available or peripherals conflict with others).
```
Comment (jamesmunns) [inline src/portability/portability.md]: * busses => buses
* there're => there are
* tradeoffs => trade-offs

**Comment by jamesmunns** [comment]
> Is this new inline suggestions GH beta feature available for you?
I don't think so, I see the old/existing online editor, but that is really the same as committing to your open branch. I can apply all changes if you like, but then I probably shouldn't be the final reviewer. I'll keep an eye out for this on future PRs, thanks for the heads up :smiley: 
> I was hoping to get feedback on the general structure and content
Definitely looks good to me. I'd say if you can add examples and they "flow" well on this page, cool, if it ends up getting too verbose, a second "guided tour" page in this section would also make sense.

**Comment by therealprof** [comment]
> I don't think so, I see the old/existing online editor, but that is really the same as committing to your open branch. I can apply all changes if you like, but then I probably shouldn't be the final reviewer. I'll keep an eye out for this on future PRs, thanks for the heads up 😃

The change suggestions are different, you can basically make suggestions (not the changes) like you did by commenting on a line but directly in the text so I can simply accept or reject it. It's a very nice feature and I've seen it in various places but still doesn't seem to be available everywhere yet: https://help.github.com/articles/incorporating-feedback-in-your-pull-request/

> Definitely looks good to me. I'd say if you can add examples and they "flow" well on this page, cool, if it ends up getting too verbose, a second "guided tour" page in this section would also make sense.

The formatting might be a bit tricky but I don't think they'd be too verbose; just a few mocked up lines to show the idea.

**Comment by therealprof** [comment]
@jamesmunns Hah, they changed the suggestions interface slightly it seems. If you click on the blue + symbol to comment on a line, there's a new symbol in the comment editor showing a document with a +/- sign in it (or use cmd-g) which will pull the line into a suggestion block. If you change there then it will be recorded as a suggestion rather than a remark and the PR author can simply add it.

**Comment by jamesmunns** [comment]
Got it, didn't notice that new icon. Thanks for pointing it out. Switched all of my comments to suggestions, other than the british vs american english stuff.

**Comment by jamesmunns** [comment]
Item 6 on [this list](https://help.github.com/articles/commenting-on-a-pull-request/#adding-line-comments-to-a-pull-request) was what I needed :)

**Comment by therealprof** [comment]
@jamesmunns Thanks a lot for trying that out with me. That's an insanely useful feature and very handy especially for documentation.

**Comment by jamesmunns** [comment]
@therealprof ok to `bors`/merge, or did you want to wait for more feedback?

**Comment by therealprof** [comment]
@jamesmunns If you're happy with it, go ahead. We can always improve with further PRs (like adding the examples).

File: src/portability/portability.md (hunk 3 above)
Comment (thejpster) [reply src/portability/portability.md]: I have a weak vote for US English, much as it annoys me personally, but a strong vote for being internally self-consistent.

**Pull request merged**